### PR TITLE
Enable kv cache hybrid allocator in TPU

### PR DESCRIPTION
**Pull request by Chenyaaang**
# Description

Enable kv cache hybrid allocator in TPU. 

This is used when a model has multiple attention types (for example: sliding window attention + full attention in gpt-oss, Gemma 2/3, Ministral, cohere), the kv cache manager needs to allocate different slots to different layers, and support layer-specific prefix-cache rules. Vllm has different types of [kv cache manager](https://github.com/vllm-project/vllm/blob/main/vllm/v1/core/single_type_kv_cache_manager.py#L245) and use different [kv cache coordinators](https://github.com/vllm-project/vllm/blob/main/vllm/v1/core/kv_cache_coordinator.py#L396) to handle this, this pr modifies `get_kv_cache_spec` and `initialize_kv_cache` logic in tpu_jax_runner to support hybrid kv cache allocator.

- Inside `get_kv_cache_spec`, create full attention spec or sliding window attention spec according to the layer's attention info. The attention info is propagated from vllm's model definition (`MODEL_IMPL_TYPE=vllm`), if model is defined in pure jax (`MODEL_IMPL_TYPE=flax_nnx`), we simply use full attention spec for each layer. We can also turn it on for pure jax, depending on whether we want to include multi-attn models in Jax.
- Inside `initialize_kv_cache`, create kv cache for each kv cache tensor, and create a map between layer_name and kv cache index to help each layer to get its corresponding kv cache. 
- Modify torchax wrapper to take in the layer_name_to_kvcache_index mapping and select the right kv cache for each layer.
- Add unit test in `test_tpu_jax_runner.py`
- The current implementation only supports sliding window attention. chunked local attention / mamba will be supported in future prs.

# Tests

- Unit tests in `test_tpu_jax_runner.py` passed.
- E2E run with one kv cache group succeeded. 
```
MODEL_IMPL_TYPE=vllm JAX_RANDOM_WEIGHTS=True  TPU_BACKEND_TYPE=jax  python examples/offline_inference.py --task=generate --model=meta-llama/Llama-3.1-8B-Instruct --max-model-len=1024 --tensor-parallel-size 1 --max-num-batched-tokens 1024 --max-num-seqs=1 --hf-config=meta-llama/Llama-3.1-8B-Instruct
```
- E2E run with multiple kv cache groups succeeded.
```
MODEL_IMPL_TYPE=vllm JAX_RANDOM_WEIGHTS=True  TPU_BACKEND_TYPE=jax  python examples/offline_inference.py --task=generate --model=google/gemma-3-27b-it --max-model-len=1024 --tensor-parallel-size 8 --max-num-batched-tokens 1024 --max-num-seqs=1 --hf-config=google/gemma-3-27b-it
```

# Perf
The baseline is from vllm commit 2a167b and tpu_commons commit 18eb3c.
- Test model (llama3-8b) with only 1 kv cache group, which isn't impacted from this change, there's no perf loss.
- Test model (gemma3-27b) with multiple kv cache groups, at long context and decode heavy use case, we can expect significant perf gain from A to B.
A: Use full attention for all layers
B: use both sliding window attention (52 layers) and full attention (10 layers)

image

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_commons/runner/jax/tpu_jax_runner.py [hunk 1]
```diff
@@ -244,51 +253,128 @@ def get_kv_cache_spec(self):
...
+                block_sizes=block_sizes,
+            )
+
     def initialize_kv_cache(self, kv_cache_config: KVCacheConfig) -> None:
         self.kv_caches: List[jax.Array] = []
+        self.maybe_reinitialize_input_batch(kv_cache_config)
+
+        kv_cache_sizes = {}
+        for kv_cache_tensor in kv_cache_config.kv_cache_tensors:
+            for layer_name in kv_cache_tensor.shared_by:
```
Comment (Chenyaaang) [inline tpu_commons/runner/jax/tpu_jax_runner.py]: I'm not super sure about here, the tensor is shared by `shared_by` layers, do they each get num_blocks equally? 
context: https://github.com/vllm-project/vllm/blob/main/vllm/v1/core/kv_cache_utils.py#L988-L999

Besides, there used to be an [assertion](https://github.com/vllm-project/vllm/blob/main/vllm/v1/worker/tpu_model_runner.py#L1645), here I think num_blocks doesn't gaurantee to be divided by len(shared_by) so I removed it.

**Comment by yaochengji** [comment]
@Chenyaaang thanks for such an important feature!

Could you also benchmark Gemma-3 model to see if we have any perf gain?

File: tpu_commons/runner/jax/tpu_jax_runner.py [hunk 2]
```diff
@@ -118,6 +121,11 @@ def __init__(
 
         # Cached draft tokens.
         self._draft_token_ids: Optional[list[list[int]]] = None
+        # Layer pairings for cross-layer KV sharing.
+        # If an Attention layer `layer_name` is in the keys of this dict, it
+        # means this layer will perform attention using the keys and values
+        # from the KV cache of `shared_kv_cache_layers[layer_name]`.
```
Comment (vanbasten23) [inline tpu_commons/runner/jax/tpu_jax_runner.py]: if key is the layer_name, what's the value? Could you add it to the comment?

Comment (Chenyaaang) [reply tpu_commons/runner/jax/tpu_jax_runner.py]: key is the layer name, value is the target layer name which it shares kv cache with.

**Comment by vanbasten23** [comment]
This is really cool! 

QQ, the test plan mentioned you used `google/gemma-3-27b-it` to test the  `multiple kv cache groups` case. I wonder how I can tell if a model requires "multiple kv cache groups". Or if it's not the model that determine that, who else determines it?

**Comment by xiangxu-google** [comment]
Thanks for adding this Chenyang! Wondering if you hve the plan to add Gemma2 and Gemma3 models to JAX.

**Comment by Chenyaaang** [comment]
> This is really cool!
> 
> QQ, the test plan mentioned you used `google/gemma-3-27b-it` to test the `multiple kv cache groups` case. I wonder how I can tell if a model requires "multiple kv cache groups". Or if it's not the model that determine that, who else determines it?

The model decides it, in this case, we use `MODEL_IMPL_TYPE=vllm`, so the model is defined in vllm/model_executor/models, and the attention info is propagated to be used in kv cache spec.

**Comment by Chenyaaang** [comment]
> Thanks for adding this Chenyang! Wondering if you hve the plan to add Gemma2 and Gemma3 models to JAX.

The current implementation can work with torchax-jax path. Probably in the future we want to add gemma into Jax, then we can refactor the attention class to record the attention type (the existing models in Jax are all using full attention only).

File: tpu_commons/runner/jax/tpu_jax_runner.py [hunk 3]
```diff
@@ -346,13 +346,23 @@ def _execute_model(
...
-                    attn_metadata,
-                    inputs_embeds,
-                )
+                if os.getenv("MODEL_IMPL_TYPE", "flax_nnx").lower() == "vllm":
+                    self.kv_caches, hidden_states = self.model_fn(
+                        self.state,
+                        self.kv_caches,
+                        input_ids,
+                        attn_metadata,
+                        tuple(self.layer_name_to_kvcache_index.items()),
```
Comment (xiangxu-google) [inline tpu_commons/runner/jax/tpu_jax_runner.py]: can you put this tuple after `input_embeds`, so that for the jax path we can skip this tuple using the `*args` wildcard.
```
    def __call__(
        self,
        kv_caches: List[jax.Array],
        input_ids: Optional[jax.Array],
        attention_metadata: AttentionMetadata,
        inputs_embeds: Optional[jax.Array] = None,
        *args,
    )
```

Then we can remove this `if else` check and run the same interface for both.

Comment (Chenyaaang) [reply tpu_commons/runner/jax/tpu_jax_runner.py]: Just did, thanks

File: tpu_commons/models/jax/model_loader.py [hunk 4]
```diff
@@ -180,6 +180,7 @@ def get_flax_model(
             hidden_states_sharding,
         ),
         donate_argnums=2,  # 0 is graphdef, 1 is state, 2 is kv_cache
+        static_argnums=6,  #6 is layer_name_to_kvcache_index.
```
Comment (yaochengji) [inline tpu_commons/models/jax/model_loader.py:183]: I would prefer `static_argnames`, WDYT? donate_argnums could also be changed to `donate_argnames`.

File: tpu_commons/models/vllm/sharding.py [hunk 5]
```diff
@@ -36,8 +38,26 @@
...
+class AttentionInfo:
+    layer_name: str
+    num_kv_heads: int
+    head_size: int
+    attn_type: str
+    sliding_window: Optional[int] = None
+    kv_sharing_target_layer_name: Optional[str] = None
+
+
 def shard_attention(layer: torch.nn.Module, mesh: Mesh,
```
Comment (yaochengji) [inline tpu_commons/models/vllm/sharding.py:51]: I'm not sure `shard_attention` is still a good name now. It's basically wrapping attention module for tpu. cc @hfan

File: tpu_commons/models/jax/model_loader.py (hunk 4 above)
Comment (Chenyaaang) [reply tpu_commons/models/jax/model_loader.py:183]: It's using *args, so we can't use argnames here.

File: tpu_commons/models/vllm/sharding.py (hunk 5 above)
Comment (Chenyaaang) [reply tpu_commons/models/vllm/sharding.py:51]: I think it's fine, it takes in `mesh` and eventually being sharded here: https://github.com/vllm-project/tpu_commons/blob/main/tpu_commons/models/jax/attention.py#L24
